### ISSUE TITLE
Next.js app uses previous commit

### ISSUE THREAD

**Issue by foxbox-doug** [aws-amplify/amplify-hosting#3131]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-hosting/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-hosting/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-hosting/blob/master/CONTRIBUTING.md).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### App Id

d2hbdzl31vlu96

### AWS Region

us-east-1

### Amplify Hosting feature

Frontend builds

### Describe the bug

When my app is updating, the previous code is still being deployed and not the latest. I have confirmed this by including the version number from package.json in __NEXT_DATA__. The only symptom I can find that indicates where the issue may be occurring is in the following warning during the Cloning Repository phase of the build:

```
2022-11-21T19:09:17.980Z [INFO]: # Cloning repository: git@github.com:[removed]/[removed].git
2022-11-21T19:09:20.414Z [INFO]: 
2022-11-21T19:09:20.416Z [INFO]: Cloning into 'removed'...
2022-11-21T19:09:20.416Z [INFO]: # Checking for Git submodules at: /codebuild/output/src132231801/src/[removed]/.gitmodules
2022-11-21T19:09:20.426Z [INFO]: # Retrieving environment cache...
2022-11-21T19:09:20.491Z [WARNING]: ! Unable to write cache: {"code":"ERR_BAD_REQUEST","message":"Request failed with status code 404"})}
2022-11-21T19:09:20.491Z [INFO]: ---- Setting Up SSM Secrets ----
2022-11-21T19:09:20.492Z [INFO]: SSM params {"Path":"/amplify/d2hbdzl31vlu96/main/","WithDecryption":true}
2022-11-21T19:09:20.553Z [INFO]: # Defaulting to Node version 16
2022-11-21T19:09:31.668Z [INFO]: # Node version 16 is available for installation
2022-11-21T19:09:31.752Z [INFO]: # Installing Node version 16
2022-11-21T19:09:44.351Z [INFO]: # Now using Node version 16
2022-11-21T19:09:44.565Z [INFO]: # Retrieving cache...
2022-11-21T19:09:44.565Z [INFO]: # Retrieved cache
```

### Expected behavior

Application should update to the latest code when RELEASE job is triggered.

### Reproduction steps

1. Create application with CloudFormation as follows:
```
AWSTemplateFormatVersion: '2010-09-09'
Description: >
  This template deploys the web application via Amplify

Parameters:
  Environment:
    Type: String
    AllowedValues:
      - dev
      - qa
      - test
      - prod
      - demo
  Index:
    Type: String
    Description: 'Numeric part of Id'
  Repository:
    Type: String
    Description: GitHub Repository
  DomainName:
    Type: String
    Description: 'Domain'
  DomainPrefix:
    Type: String
    Description: 'Subdomain Prefix'
    Default: 'www'
  Branch:
    Type: String
    Description: Branch to deploy
    Default: main
  AccessToken:
    Type: String
    Description: GitHub Personal Access Token (needs admin:repo_hook)
  UseQA:
    Type: String
    Description: Set to true to use qa settings
    AllowedValues:
      - true
      - false
    Default: 'false'

Mappings:
  EnvironmentMap:
    Stage:
      dev: DEVELOPMENT
      qa: BETA
      test: BETA
      demo: PRODUCTION
      prod: PRODUCTION

Conditions:
  UseQA:
    Fn::Equals: [ !Ref UseQA, 'true' ]

Resources:
  AmplifyApp:
    Type: AWS::Amplify::App
    Properties:
      AccessToken: !Ref AccessToken
      BuildSpec: |
        version: 1
        applications:
          - frontend:
              phases:
                preBuild:
                  commands:
                    - cd ../.. & yarn install
                    - echo "NEXT_PUBLIC_CLIENT_ID=${NEXT_PUBLIC_CLIENT_ID}" >> .env
                    - echo "NEXT_PUBLIC_COGNITO_URL=${NEXT_PUBLIC_COGNITO_URL}" >> .env
                    - echo "NEXT_PUBLIC_BASE_URL=${NEXT_PUBLIC_BASE_URL}" >> .env
                    - echo "NEXT_PUBLIC_CONTENT_URL=${NEXT_PUBLIC_CONTENT_URL}" >> .env
                build:
                  commands:
                    - yarn build
              artifacts:
                baseDirectory: .next
                files:
                  - '**/*'
              cache:
                paths:
                  - node_modules/**/*
            appRoot: packages/subscriber
      Description: !Sub 'Amplify application for ${Environment}-${Index}'
      EnvironmentVariables:
        - Name: AMPLIFY_DIFF_DEPLOY
          Value: false
        - Name: AMPLIFY_MONOREPO_APP_ROOT
          Value: packages/subscriber
        - Name: NODE_OPTIONS
          Value: --max_old_space_size=4096
        - Name: _LIVE_UPDATES
          Value: '[{"pkg":"next-version","type":"internal","version":"latest"}]'
      IAMServiceRole: !Sub arn:aws:iam::${AWS::AccountId}:role/AmplifyRole
      Name: !Sub '${Environment}-${Index}'
      Repository: !Ref Repository
  AmplifyBranch:
    Type: AWS::Amplify::Branch
    Properties:
      AppId: !GetAtt  AmplifyApp.AppId
      BranchName: !Ref Branch
      Description: !Sub 'Amplify deployment for ${Environment}-${CreatorIndex} master'
      EnableAutoBuild: false
      EnvironmentVariables:
        - Name: NEXT_PUBLIC_CLIENT_ID
          Value:
            Fn::If:
              - UseQA
              - {'Fn::ImportValue': !Sub 'qa-UserPoolClient'}
              - {'Fn::ImportValue': !Sub 'serverlessrepo-${Environment}-${Index}-UserPoolClient'}
        - Name: NEXT_PUBLIC_COGNITO_URL
          Value:
            Fn::If:
              - UseQA
              - !Sub 'https://qa-${AWS::AccountId}.auth.${AWS::Region}.amazoncognito.com'
              - !Sub 'https://${Environment}-${Index}-${AWS::AccountId}.auth.${AWS::Region}.amazoncognito.com'
        - Name: NEXT_PUBLIC_BASE_URL
          Value:
            Fn::If:
              - UseQA
              - {'Fn::ImportValue': !Sub 'qa-1-creator-WebEndpoint'}
              - {'Fn::ImportValue': !Sub 'serverlessrepo-${Environment}-${CreatorIndex}-WebEndpoint'}
        - Name: NEXT_PUBLIC_CONTENT_URL
          Value:
            Fn::If:
              - UseQA
              - {'Fn::ImportValue': !Sub 'qa-1-creator-ContentURL'}
              - {'Fn::ImportValue': !Sub 'serverlessrepo-${Environment}-${CreatorIndex}-ContentURL'}
      Stage: !FindInMap [ EnvironmentMap, Stage, !Ref Environment ]
  DomainRecord:
    Type: AWS::Amplify::Domain
    Properties:
      AppId: !GetAtt AmplifyApp.AppId
      DomainName: !Ref DomainName
      SubDomainSettings:
        - Prefix: !Ref DomainPrefix
          BranchName: !GetAtt AmplifyBranch.BranchName

Outputs:
  URL:
    Description: Application URL
    Value: !Sub 'https://${Branch}.${AmplifyApp.DefaultDomain}'
  AppId:
    Description: Application ID
    Value: !GetAtt AmplifyApp.AppId
    Export:
      Name: !Sub '${AWS::StackName}-AppId'
  Branch:
    Description: Branch Name
    Value: !Ref Branch
    Export:
      Name: !Sub '${AWS::StackName}-Branch'
```
2. Set main branch via UI (this appears to not be set when creating app via CloudFormation)
3. Push commit to GitHub
4. Trigger RELEASE job via `aws amplify start-job --branch-name main --app-id d2hbdzl31vlu96  --job-type RELEASE`
5. Repeat 3-4 until app no longer appears to update

### Build Settings

```yaml
version: 1
applications:
  - frontend:
      phases:
        preBuild:
          commands:
            - cd ../.. & yarn install
            - echo "NEXT_PUBLIC_CLIENT_ID=${NEXT_PUBLIC_CLIENT_ID}" >> .env
            - echo "NEXT_PUBLIC_COGNITO_URL=${NEXT_PUBLIC_COGNITO_URL}" >> .env
            - echo "NEXT_PUBLIC_BASE_URL=${NEXT_PUBLIC_BASE_URL}" >> .env
            - echo "NEXT_PUBLIC_CONTENT_URL=${NEXT_PUBLIC_CONTENT_URL}" >> .env
        build:
          commands:
            - yarn build
      artifacts:
        baseDirectory: .next
        files:
          - '**/*'
    appRoot: packages/subscriber
```

### Log output

<details>

```
# Put your logs below this line
2022-11-21T19:09:17.980Z [INFO]: # Cloning repository: git@github.com:[removed]/[removed].git
2022-11-21T19:09:20.414Z [INFO]: 
2022-11-21T19:09:20.416Z [INFO]: Cloning into 'removed'...
2022-11-21T19:09:20.416Z [INFO]: # Checking for Git submodules at: /codebuild/output/src132231801/src/[removed]/.gitmodules
2022-11-21T19:09:20.426Z [INFO]: # Retrieving environment cache...
2022-11-21T19:09:20.491Z [WARNING]: ! Unable to write cache: {"code":"ERR_BAD_REQUEST","message":"Request failed with status code 404"})}
2022-11-21T19:09:20.491Z [INFO]: ---- Setting Up SSM Secrets ----
2022-11-21T19:09:20.492Z [INFO]: SSM params {"Path":"/amplify/d2hbdzl31vlu96/main/","WithDecryption":true}
2022-11-21T19:09:20.553Z [INFO]: # Defaulting to Node version 16
2022-11-21T19:09:31.668Z [INFO]: # Node version 16 is available for installation
2022-11-21T19:09:31.752Z [INFO]: # Installing Node version 16
2022-11-21T19:09:44.351Z [INFO]: # Now using Node version 16
2022-11-21T19:09:44.565Z [INFO]: # Retrieving cache...
2022-11-21T19:09:44.565Z [INFO]: # Retrieved cache

```

</details>

**Comment by hloriana** [comment]
Hi @foxbox-doug 👋🏽 we apologize for this inconvenience. Thank you for providing thorough reproduction steps. We will investigate this behavior and report back with next steps for a possible work around.

**Comment by hloriana** [comment]
Upon further investigation, this can be the result of a race condition. If a new job is starting while a previous job is still running, then it is possible that the previous job will be deployed while the newer job fails, causing the previous commit to be deployed instead of the new commit. You can prevent this by canceling any currently running builds before starting a new job:

`aws amplify stop-job --app-id <value> --branch-name <value> --job-id <value>`

Please let us know if this race condition is not applicable to your scenario and we will continue to investigate.

**Comment by foxbox-doug** [comment]
> Upon further investigation, this can be the result of a race condition. If a new job is starting while a previous job is still running, then it is possible that the previous job will be deployed while the newer job fails, causing the previous commit to be deployed instead of the new commit. You can prevent this by canceling any currently running builds before starting a new job:
> 
> `aws amplify stop-job --app-id <value> --branch-name <value> --job-id <value>`
> 
> Please let us know if this race condition is not applicable to your scenario and we will continue to investigate.

Definitely not race condition. Waited for all previous jobs to complete before queuing new job.

**Comment by hloriana** [comment]
@foxbox-doug it appears you have deleted this application. Please let us know if you still need assistance with this issue.

**Comment by foxbox-doug** [comment]
@hloriana We had to recreate the app to get it to update. I will provide more info if the issue repeats itself.

**Comment by GabrielArakaki** [comment]
Hi everyone! I don't have a functional evidence here, but I found the same issue when I migrated my application to web compute platform through the wizard.

![image](https://user-images.githubusercontent.com/7118128/203799733-1cbb1a35-c8ab-404d-bd13-eaf1c1fb2483.png)

My feature branches started to use another unrelated branch, and was actually using a previous commit before HEAD.
My solution was to create a whole new app in web compute and it worked from the start.

**Comment by hloriana** [comment]
Hi @GabrielArakaki 👋🏽 thank you for providing details around this behavior. Do you happen to have the App ID of the application that was migrated? This could potentially help us investigate this behavior further.

**Comment by GabrielArakaki** [comment]
Hi @hloriana Sorry, I have also deleted the old application!

**Comment by edenriquez** [comment]
is there any update about sites not reflecting changes after migration ? @hloriana

**Comment by kivohin** [comment]
Hi! 🙂

We are experiencing the same issue, that is even though we successfully push and build new code changes, a previous/old commit is being deployed.

This started happening after pressing the "Migrate" button, which came after the announcement of Next.js 12 and 13 support in the new web compute platform.

Next.js version 12.3.4
Node.js 16.18.1
App ID d7rf8fcnzpljj
Region us-east-1

Please help, thank you.
CC: @hloriana

**Comment by foobarnes** [comment]
@edenriquez @kivohin It looks like Amplify Hosting now directly manages the resources that were previously deployed to CloudFront, Lambda, etc.

My issue involved rewrite rules, which pointed to old distributions, not being automatically removed. 
[See answer here](https://github.com/aws-amplify/amplify-hosting/issues/3158#issuecomment-1332755075)

**Comment by kivohin** [comment]
Thank you @foobarnes! Indeed, that was one of the issues for us as well, i.e. we had some old rewrite rules pointing to old CloudFront distributions.

Our other issue was that we manage our custom domain name configurations through Route53, and not through Amplify's `Domain management` section.

Since the Amplify Hosting model changed, and they are now directly managing all the resources, we are now forced to setup our custom domain through Amplify's `Domain management` section.

After doing that everything came back to normal.

Definitely an interesting change Amplify Hosting is heading into. More into the Vercel/Netlify direction where they have the control over everything.

**Comment by zevi-wiply** [comment]
Hi, we are still experiencing same issue of sites not reflecting changes post-migration even after removing old rewrite rules.
@foobarnes @hloriana any ideas?

App ID is d9vnksjwvp8m7

Builds are successful and the frontend seems to be in the right commit? But it is definitely an older one that is getting deployed.

**Comment by drmzio** [comment]
Regarding @zevi-wiply's issue, this is also happening to my app as well. Started happening after switching to Web compute.

**Comment by uokesita** [comment]
We are experiencing this issue also.
We have a NextJS app deployed to an amplify environment.

When using NextJS version `12.1.8` and `WEB_DYNAMIC` it works perfectly.
When using NextJS version `13.0.6` and `WEB_COMPUTE ` it fails to display the latest commit.

**Comment by uokesita** [comment]
After some investigation, this is what we did to solve the issue:

- Upgrading to NextJS 13, requires changing the platform to `WEB_COMPUTE`.
  Moving to `WEB_COMPUTE`, requires deleting the rewrite rules as they are no longer needed, maintaining the rules, redirects to the **last successful deploy** made with `WEB_DYNAMIC`.
  This is a configuration on the Amplify dashboard that needs to be removed manually.

- Using `WEB_COMPUTE` expects the build to export a standalone app (server and frontend).
  We needed to configure NextJS for generating a build to be a standalone app by doing in `nextjs.config`:

    ```
  output: 'standalone',
  experimental: {
      outputFileTracingRoot: path.join(__dirname, '../../'),
  },
    ```


- Using `WEB_COMPUTE` and, a **Monorepo**, expects the server output to be located in other folder different from the exported standalone default.
  We had to configure the Amplify build script to move the assets to the expected location so, Amplify can locate and start the server, by doing in `amplify.yml`:
  
    ```
    postBuild:
      commands:
        - echo FRONTEND POSTBUILD
        - shopt -s dotglob
        - mv -f .next/standalone/app/app-name/node_modules/* .next/standalone/node_modules
        - rm -rf .next/standalone/app/app-name/node_modules
        - mv -f .next/standalone/app/app-name/* .next/standalone
        - rm -rf .next/standalone/app/app-name
   ```

Im not sure if this is the correct thing to do, but I hope this helps someone or the Amplify team

**Comment by hloriii** [comment]
At the time this issue was opened, Amplify was failing to remove the redirect rule that points to the previous CloudFront distribution after migrating the application. This has since been resolved and should no longer be an issue for folks who need to migrate to `WEB_COMPUTE`.
